lib/thread.c

```diff
@@ -1669,10 +1732,48 @@ void thread_call(struct thread *thread)
 	unsigned long helper;
 #endif
 	RUSAGE_T before, after;
+	struct timeval wkptime, diff;
+	uint64_t wkpintvl_usecs;
 
 	GETRUSAGE(&before);
 	thread->real = before.real;
 
+	if (thread_collect_stats) {
+		/*
+		 * The following code measures interval between consecutive
```

Comment:
I've read this comment, but I don't see how this computation actually helps much. you're measuring how often tasks associated with this pthread get scheduled, is that right? I don't see what that actually indicates. why not just measure overall cpu utilization, for example - if that is "high" (whatever "high" means) then some processes are going to be starved to some extent?

Reply:
@mjstapp The purpose is to NOT detect CPU hogs caused by the current process. It is actually to detect if the current porocess and its thread has been CPU starved or not (possibly due to CPU hogging from another process, not necessarily FRR daemon). So we log the maximum interval between two consecutive wakeup of the same thread.

Reply:
you're measuring the time between each successive pair of tasks. tasks that do a lot of work might take a long time, and tasks that do little work will take a short time - isn't there a lot of variability there? what does the number mean? if there's something going on in the environment, in the overall system hosting the frr process, will that be detected at microsecond granularity? I don't think so, I guess - that fine granularity isn't going to tell you anything about overall system load.
I can imagine injecting a 1-second timer, say, and examining how close to the 1-second interval it's getting scheduled. that can still be perturbed by the scheduling pattern of a busy daemon, and doesn't necessarily mean that some external process is the source of latency? the only way to tell whether the overall system is busy is to look at the overall system cpu utilization, isn't it?